src/core/lib/event_engine/windows/windows_endpoint.cc

```diff
@@ -0,0 +1,333 @@
+  DWORD flags = 0;
+  // First let's try a synchronous, non-blocking read.
+  int status = WSARecv(socket_->socket(), wsa_buffers, (DWORD)buffer->Count(),
+                       &bytes_read, &flags, nullptr, nullptr);
+  int wsa_error = status == 0 ? 0 : WSAGetLastError();
+  // Did we get data immediately ? Yay.
+  if (wsa_error != WSAEWOULDBLOCK) {
+    // prune slicebuffer
+    if (bytes_read != buffer->Length()) {
+      buffer->RemoveLastNBytes(buffer->Length() - bytes_read);
+    }
+    executor_->Run([on_read = std::move(on_read)]() mutable {
+      on_read(absl::OkStatus());
+    });
+    return;
+  }
+  // Otherwise, let's retry, by queuing a read.
+  memset(socket_->read_info()->overlapped(), 0, sizeof(OVERLAPPED));
+  status =
+      WSARecv(socket_->socket(), wsa_buffers, (DWORD)buffer->Count(),
+              &bytes_read, &flags, socket_->read_info()->overlapped(), nullptr);
+  wsa_error = status == 0 ? 0 : WSAGetLastError();
+  if (wsa_error != 0 && wsa_error != WSA_IO_PENDING) {
+    // Async read returned immediately with an error
+    executor_->Run([this, on_read = std::move(on_read), wsa_error]() mutable {
+      on_read(WSAErrorToStatusWithMessage(
+          wsa_error, absl::StrFormat("WindowsEndpont::%p Read failed", this),
+          DEBUG_LOCATION));
+    });
+    return;
+  }
+
+  handle_read_event_.SetCallback(std::move(on_read));
+  handle_read_event_.SetSliceBuffer(buffer);
+  socket_->NotifyOnRead(&handle_read_event_);
+}
+
+void WindowsEndpoint::Write(absl::AnyInvocable<void(absl::Status)> on_writable,
+                            SliceBuffer* data, const WriteArgs* /* args */) {
+  if (grpc_event_engine_trace.enabled()) {
+    for (int i = 0; i < data->Count(); i++) {
+      auto str = data->RefSlice(i).as_string_view();
+      gpr_log(GPR_INFO, "WindowsEndpoint::%p WRITE (peer=%s): %.*s", this,
+              peer_address_string_.c_str(), str.length(), str.data());
+    }
+  }
+  GPR_ASSERT(data->Count() <= UINT_MAX);
+  WSABUF local_buffers[kMaxWSABUFCount];
+  LPWSABUF buffers = local_buffers;
+  LPWSABUF allocated = nullptr;
+  MemoryAllocator::Reservation memory_reservation;
+  if (data->Count() > kMaxWSABUFCount) {
+    size_t bytes_to_allocate = sizeof(WSABUF) * data->Count();
+    memory_reservation = allocator_.MakeReservation(bytes_to_allocate);
+    allocated = (WSABUF*)gpr_malloc(bytes_to_allocate);
+    buffers = allocated;
+  }
+  for (int i = 0; i < data->Count(); i++) {
+    auto slice = data->RefSlice(i);
+    GPR_ASSERT(slice.size() <= ULONG_MAX);
+    buffers[i].len = slice.size();
+    buffers[i].buf = (char*)slice.begin();
+  }
+  // First, let's try a synchronous, non-blocking write.
+  DWORD bytes_sent;
+  int status = WSASend(socket_->socket(), buffers, (DWORD)data->Count(),
+                       &bytes_sent, 0, nullptr, nullptr);
+  size_t async_buffers_offset;
+  if (status == 0) {
+    if (bytes_sent == data->Length()) {
+      // Write completed, exiting early
+      executor_->Run(
```

Comment:
Interesting question I don't know the answer to: we do a synchronous write for efficiency here, but then need an extra allocation to call the callback.... does this render the synchronous path moot?